### ISSUE TITLE
Add a `TouchActionsPanel` to Android Editor

### ISSUE THREAD

**Issue by syntaxerror247** [godotengine/godot-proposals#11342]
- Related: https://github.com/godotengine/godot-proposals/issues/11196

### Describe the project you are working on

Godot Android Editor

### Describe the problem or limitation you are having in your project

The Godot editor relies heavily on keyboard shortcuts for Undo (Ctrl+Z) and Redo (Ctrl+Shift+Z) actions. However, on touch-based devices without a keyboard, accessing these actions becomes cumbersome, requiring users to navigate menus that disrupt the workflow. This limitation reduces productivity and makes the editor less convenient for Android users.

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

This proposal introduces a floating `Touch Actions Panel` node to the Android editor interface. The panel includes common shortcuts like Undo and Redo buttons designed for touch-based workflows. Positioned initially at the bottom-center of the screen, the panel is draggable, allowing users to reposition it according to their preferences.

The _floating nature_ of the panel improves usability by providing quick and intuitive access to essential actions without requiring a physical keyboard or navigating complex menus. This enhancement caters specifically to touch-first devices, making the editor more accessible and user-friendly for Android users.

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

Implementation Details:

- The panel is a TouchActionsPanel node that can be freely repositioned by dragging a drag handle located on its left side.
- Initially, the panel is anchored to the bottom-center of the editor interface.
- The panel uses a rounded, translucent background to blend seamlessly with the editor's UI.
- The buttons trigger editor shortcut actions like ui_undo or ui_redo.
- Drag functionality is implemented to allow smooth repositioning of the panel.

**Future Expandability**: The panel is designed to be extensible, allowing additional buttons or actions to be added. For example, buttons for Save, Cut, Copy, and Paste could be added.

https://github.com/user-attachments/assets/0617504f-e557-4685-86b9-e35a7e5b4e59

### If this enhancement will not be used often, can it be worked around with a few lines of script?

While the enhancement could be implemented as an editor plugin, doing so would require manual installation and maintenance, making it less accessible to the broader user base. A core implementation ensures seamless integration into the Android editor experience.

### Is there a reason why this should be core and not an add-on in the asset library?

This is about enhancing the _usability_ of the Android Editor. Shortcuts like Undo and Redo are fundamental functionalities essential to any editing workflow. Providing touch-friendly access to these actions ensures a seamless and efficient user experience on Android devices. Integrating this feature into the core editor avoids fragmentation, guarantees consistency, and ensures that all Android users can benefit from this improvement without requiring additional setup or plugins.

**Comment by tetrapod00** [comment]
One thing to consider with your proposed UI is how it works when the bottom panel is full of tabs, which can already be a problem. Are your buttons added to the bottom panel tab HBox, between the tabs and the buttons on the right? Or are they rendered on top?

---

Side note, I wonder why the default desktop editor doesn't have dedicated Undo and Redo buttons, either. I looked briefly for a proposal or PR to add them and couldn't find one. So it doesn't seem to be a case where it was discussed and declined; it was just never discussed?

One issue with adding an undo and redo button (to the desktop editor) is where it would go. You can't really put it in the toolbar, because the toolbar is workspace-specific. The `History` dock makes sense but then it's usually inaccessible. There is space between the workspace selectors and the Run Project, etc buttons.

I guess it's just not something that really needs dedicated buttons anymore, in modern desktop UIs.

**Comment by syntaxerror247** [comment]
> Are your buttons added to the bottom panel tab HBox, between the tabs and the buttons on the right? Or are they rendered on top?

Currently, It is between the tabs and the buttons on the right. Though, I'm thinking about making it a floating panel.

> Side note, I wonder why the default desktop editor doesn't have dedicated Undo and Redo buttons, either.

Undo/Redo buttons are in `Scene` menu.

**Comment by syntaxerror247** [comment]
I have updated this proposal to reflect the latest changes. The feature now includes a **floating Touch Actions Panel** with Save, Undo, and Redo buttons, along with a draggable handle for repositioning. This enhancement replaces the earlier plan to place the buttons in the bottom toolbar.

@m4gr3d I believe you might find this feature interesting for the Android editor. Let me know your thoughts!

**Comment by llama-nl** [comment]
@syntaxerror247 I don't think that we'll need extra undo/redo button. Cause there is history menu next to the single menu where all the history available. You can make it customizable like if any node selected delete button apear. 

And you can fix the action UI vertically with only icons on left top side of the viewport like other 3d engine for Android.

**Comment by syntaxerror247** [comment]
@llama-nl Thanks for your suggestions!

This is an initial implementation idea. I plan to add more buttons and refine the UI as it progresses.

Regarding the undo/redo buttons, I believe the history tab doesn't work with the script editor. I also believe dedicated buttons can offer a faster and more intuitive way to undo or redo actions.

As for the vertical layout, it's already planned. The panel will include a button to toggle between vertical and horizontal layouts. However, I'm not in favor of a fixed layout. While a fixed layout might suit some users, others may prefer a more flexible approach. I'm planning to introduce an option to pin the panel, preventing it from being dragged.

**Comment by llama-nl** [comment]
Oh I understand. Let me know, when you'll open PR. I am so excited to test the next level implemention. 😃

**Comment by m4gr3d** [comment]
> I have updated this proposal to reflect the latest changes. The feature now includes a **floating Touch Actions Panel** with Save, Undo, and Redo buttons, along with a draggable handle for repositioning. This enhancement replaces the earlier plan to place the buttons in the bottom toolbar.
> 
> [@m4gr3d](https://github.com/m4gr3d) I believe you might find this feature interesting for the Android editor. Let me know your thoughts!

@syntaxerror247 Sounds like a good idea! Are you thinking of the panel being implemented as an Android view, or using Godot UI nodes?

**Comment by m4gr3d** [comment]
> This is an initial implementation idea. I plan to add more buttons and refine the UI as it progresses.

@syntaxerror247 Might be worth taking a look at Xogot's UI as Miguel did a great job providing touch friendly UI affordances.

https://youtu.be/irVRaTj0SGU

**Comment by m4gr3d** [comment]
> > This is an initial implementation idea. I plan to add more buttons and refine the UI as it progresses.
> 
> [@syntaxerror247](https://github.com/syntaxerror247) Might be worth taking a look at Xogot's UI as Miguel did a great job providing touch friendly UI affordances.
> 
> https://youtu.be/irVRaTj0SGU

cc @Calinou @fire there was a discussion in the `#editor-design` channel about bringing into core some of the improved UI Miguel created for Xogot; might be a good idea to use this proposal as a first step in that direction!

**Comment by GDevSh007** [comment]
@syntaxerror247  add another button to it in 4.7 as 4.6 has already entered feature freeze. Suggestion is an addition of another button to toggle the virtual keyboard on and off as it bothers quite a lot when editing scripts.
